Feature: Ability to hide selected toolbar buttons via xml config

ThioJoe opened the pull request:
I ended up taking a crack at an issue that has been bothering me for a while, which is the overwhelming amount of toolbar buttons, especially as more plugins get installed.

Basically this addition creates a new xml config file in the AppData folder which lists all the button commands for each plugin, as well as the built in buttons, and a 'hide' attribute you can set to hide that button. You can also hide all buttons and override the individual hide settings for a particular plugin by setting the respective `hideAll` attributes to yes.

Corresponding feature request:
resolves #16000 

It could partly address #15106 (though that person seems to want more advanced features like combining plugin buttons. Apparently the Customize Toolbar plugin has compatibility issues with recent versions of Notepad++ as discussed in #15440, so this feature could be a replacement for those who simply want to hide / show certain buttons.

### Example - Before (Default everything showing):

<img width="480" alt="image" src="https://github.com/user-attachments/assets/f18146ad-7ae0-48f9-8e27-afe1020713fa" />


### Example - After: First several built in buttons hidden:

<img width="480" alt="image" src="https://github.com/user-attachments/assets/abfc9602-3087-4243-a7e0-81a6e9accaf3" />


### Details

Ultimately the way it hides buttons is within `ToolBar::init` inside `Toolbar.cpp`, where it checks if the button about to be processed/added is one that should be hidden, and if so it adds the `TBSTATE_HIDDEN` flag to the `_pTBB[i].fsState` value that gets set. 

Importantly this means behind the scenes the button is still there, just invisible, so it shouldn't mess up other references to that button or something.

The actual bulk of the logic is in a single function in `Toolbar.cpp` called `initToolbarButtonVisibilityConfig`. It uses two new structs, `CommandInfo` and `ToolbarButtonConfig` which I used to collect all the necessary info about each button in one place. 

<img width=50% alt="image" src="https://github.com/user-attachments/assets/e9e5cb07-6641-4dad-adfa-d51004c3599d" />

There's also a struct that's just basically serves as a sort of enum for string values mostly used with the XML file since hardcoding strings in the logic can be a pain I think

<img width="431" alt="image" src="https://github.com/user-attachments/assets/eb45eec3-d285-4cb4-ac34-baa46b1accc9" />

------

`initToolbarButtonVisibilityConfig` is a big function but its very linear so figured it was better to keep it all together. The basics of the logic are:
1.  Get the info about all the built-in commands and plugin commands, and determine which are for buttons
2. Load the XML config file (if there is one)
3. Match up the user setting for each button using the name of the command, and apply it to the `ToolbarButtonConfig` object.
4. Using dedicated function `writeToolbarButtonsConfig`, write an updated or entirely new version of the XML file if list of commands between the xml file and the initially obtained button command list is different. It will still match up whichever command settings it can even if others are missing or new, and retain those settings when updating the xml file.

`initToolbarButtonVisibilityConfig`  gets called within `Notepad_plus.cpp` right before the `init` function. Then in there is some basic logic for checking if the command ID matches one that's hidden in the config object. Plus a bit extra logic to prevent consecutive separators if many of the built-in functions are hidden.

------


### Stuff I wasn't sure about
- I just put the name of the file as a constant at the top of Toolbar.cpp, but I figure there might be a better place for it like with other config files, but couldn't really find it
- I couldn't find an existing constant defined anywhere for the separator command ID that seems to be 0, so I just made one `SEPARATOR_CMD_ID` in `ToolBar.h` since the guidelines mention avoiding using 'magic numbers', and that command ID of 0 for separators appeared in several places even though I guess it's a placeholder anyway.

Overall this is a pretty basic implementation of the feature, though it does work, and I am not aware of any issues (it's highly possible I missed something though). I'm not familiar with C++ at all so you might want to double check the parts dealing with reading/writing the XML files since apparently TinyXML requires the usage of pointers as far as I can tell, so I'm not sure if I handled all that correctly.

Anyway even if you decide not to add it as-is, maybe some of it can be used as scaffolding in the future or something.

alankilborn commented:
>I ended up taking a crack at an issue that has been bothering me for a while

Is there an actual "issue" for this?
The protocol is to open an "issue" to solicit comments from others on the appropriateness of a change.

ThioJoe commented:
> >I ended up taking a crack at an issue that has been bothering me for a while
> 
> 
> 
> Is there an actual "issue" for this?
> 
> The protocol is to open an "issue" to solicit comments from others on the appropriateness of a change.

No because I didn't realize about the guidelines until it was mostly done, definitely my bad there 💀

I can create one though to gauge any feedback. I have to imagine it would be welcomed.

donho reviewed (changes requested):
**Load and apply toolbar button config XML file, to update command infos with user settings** part should be between `// toolbarIcons.xml : for per user //` and `// shortcuts.xml : for per user //`:
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/Parameters.cpp#L1498

"toolbarButtonVisibility.xml" should be optional - it means if it's absence, then the old code (current code in master) should be run. But when this file is present & loaded successfully, then the new code (your modification) should be called to hide icons.

Don't write "toolbarButtonVisibility.xml" automatically, but instead, provide "toolbarButtonVisibility_exemple.xml" to user for the modification in case of need.


In "toolbarButtonVisibility_exemple.xml", use_ `<Button Show="no" Name="New" />` instead of `<Button Hide="yes" Name="New" />`

alankilborn commented:
I think that hiding toolbar buttons (via capability in native N++) opens the door wider for users wanting full customization (new toolbar buttons, new bitmaps, etc).  This is not a bad thing, but it is not trivial to implement, i.e., it would be a lot of work.

It would also make the _Customize Toolbar_ obsolete, which is probably a good thing, as it is buggy and a bit quirky to use.

donho commented:
@alankilborn 
> I think that hiding toolbar buttons (via capability in native N++) opens the door wider for users wanting full customization (new toolbar buttons, new bitmaps, etc). 

By new toolbar buttons I think you mean new "commands"?
We offer users the stuffs if it's possible. But regarding adding new commands - it'll be much harder to be implement, so it won't be done.

alankilborn commented:
>By new toolbar buttons I think you mean new "commands"?

Well, yes.  For example, with _Customize Toolbar_ plugin it is possible to tie a macro or a plugin command to a new button.

>it'll be much harder to be implement, so it won't be done.

Yes, I'm just thinking that if you give users ability to remove toolbar buttons, the first thing next they'll ask for is adding new buttons.

ThioJoe commented:
Ok I've moved the XML loading to its own dedicated static function for better modularity and put that call in NppParameters like you suggested.

For changing the property in the config from 'Hide' to 'Show', that's actually what I had thought to do originally. But had changed it to be 'hide' for a couple reasons:
 - With it as 'hide', it remains more consistent with the 'HideAll' option. Otherwise it might be confusing or seem contradictory to have 'show=yes' for individual settings but 'hideAll' also as yes.
 - With it as 'show', it might imply that the entry for that command must be included in the xml file (and it be set to 'yes') for it to show. But in reality, any button entries not included in the config just show as default. 

Let me know your thoughts about that.

-----

Next for the example file, I've implemented it but I realize it might be different than what you were imagining. I just realized that the other example files are provided as static file resources. 

What I've done here is made it auto-generate the example file if it does not exist, and it includes entries for all the commands (built in and plugins) so the user doesn't need to add them for all their plugins manually which would be very tedious. But it does have logic to only perform read/write actions on the xml stuff if necessary. Specifically
- It will only create the `toolbarButtonVisibility_example.xml` xml file if it doesn't exist already
- It will only gather info about the toolbar commands if necessary, specifically either:
   - The user either has created `toolbarButtonVisibility.xml` (non-example version), 
   - OR the example file needs to be created, in which case it will gather the data to be able to auto create that with all the updated command info.
     - So for clarity: Even if the user installs new plugins or removes one, the example file will not be touched unless the user deletes the previous one. This is so we don't need to gather all the data in `initToolbarButtonVisibilityConfig()` to compare it, we can just skip all the gathering altogether.

Therefore if the user has not created a toolbar customization xml, and the example xml already exists, almost all new code will be bypassed, meaning users who choose not to use the feature will not be impacted by any startup performance hit no matter how small.

ThioJoe commented on PowerEditor/src/WinControls/ToolBar/ToolBar.cpp:
```diff
@@ -122,7 +124,375 @@ void ToolBar::initTheme(TiXmlDocument *toolIconsDocRoot)
+
+	// -----------------------------------------------------------------------------------------------
+	// Load and apply toolbar button config XML file, to update command infos with user settings
+	// -----------------------------------------------------------------------------------------------
```
Yea it should already just run as normal if the xml file failed to load, it would just do the same thing as it does on the first run when the xml file doesn't exist before it created it. Then within `init()` in Toolbar.cpp it still defaults to the buttons being visible unless there's a commandInfos entry that marks it as hidden in this line:
`for (const CommandInfo& command : toolbarButtonConfig._commandInfos)`
So if commandInfos is empty for some reason it would just show the button as usual.

In any case I've further updated it to have even less impact if the user doesn't have the xml file. I've described the new changes in more detail in my latest response in this thread.

donho commented on PowerEditor/src/WinControls/ToolBar/ToolBar.cpp:
```diff
@@ -26,6 +26,8 @@ using namespace std;
 
 constexpr DWORD WS_TOOLBARSTYLE = WS_CHILD | WS_VISIBLE | WS_CLIPCHILDREN | WS_CLIPSIBLINGS | TBSTYLE_TOOLTIPS | TBSTYLE_FLAT | CCS_TOP | CCS_NOPARENTALIGN | CCS_NORESIZE | CCS_NODIVIDER;
 
+const std::wstring TOOLBAR_CONFIG_EXAMPLE_FILE = L"toolbarButtonVisibility_example.xml";
```
By asking to have "toolbarButtonVisibility_example.xml", I mean that this file should be provide as (in the same directory):
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/master/PowerEditor/src/tabContextMenu_example.xml

and it will be installed by Notepad++ installer, so if user want to hide any icon, they can modify the content of "toolbarButtonVisibility_example.xml", and rename it to "toolbarButtonVisibility.xml", then restart Notepad++.

In this PR, only the name "toolbarButtonVisibility.xml" should be used - so if it exists, it will be loaded. If the load of "toolbarButtonVisibility.xml" is successful, it'll be processed, otherwise, it's ignored completely.

donho reviewed (changes requested):
Please consider the 2 points, and inspired by loading & using tabContextMenu.xml to modify your PR.
